docs: how to create a GitHub Action (#1196)
JoshuaKGoldberg/create-typescript-app: docs/FAQs.md

Reviewer: Moving @johnnyreilly's https://github.com/JoshuaKGoldberg/create-typescript-app/pull/1196#issuecomment-1894349808 here:

> where do things stand with this?

I think I'm still waiting on resolution or re-pings on the open threads? I don't remember having any action items on my end.

Author: Think it's all resolved?

   - delete `.github/workflows/release.yml` and `.github/workflows/post-release.yml`.
   - update `.github/workflows/build.yml` to ensure `dist` is up to date:

     <details>
         <summary><code>.github/workflows/build.yml</code></summary>

     ```yml
     jobs:
       build:
         runs-on: ubuntu-latest
         steps:
           - uses: actions/checkout@v4
           - uses: ./.github/actions/prepare
           - run: pnpm build

           - name: Compare dist/index.js
             run: |
               if [ "$(git diff --ignore-space-at-eol --text dist/index.js | wc -l)" -gt "0" ]; then
                 echo "Detected uncommitted changes after build."
                 echo "You may need to run 'pnpm run build' locally and commit the changes."
                 echo ""
                 echo "See diff below:"
                 echo ""
                 git diff --ignore-space-at-eol --text dist/index.js
                 echo ""
                 # say this again in case the diff is long
                 echo "You may need to run 'pnpm run build' locally and commit the changes."
                 echo ""
                 exit 1
               fi

> 
>            - id: upload
>              if: ${{ failure() && steps.diff.outcome == 'failure' }}
>              name: Upload artifact
>              uses: actions/upload-artifact@v4
>              with:
>                name: dist
>                path: dist/

Reviewer: Similar to #1225: if we don't need an artifact, I don't think we should add config complexity to wrangle it.

```suggestion
```

In this case I think it should be enough to print the diff out and suggest running `pnpm run build`.

         branches:
           - main

     permissions:
       contents: read
     ```

      </details>

   - GitHub Actions run without installing package dependencies.
     Replace `tsup` with [`ncc`](https://github.com/vercel/ncc) to build source files and dependencies into a single JS file.
     Delete `tsup.config.ts` then execute the following commands:

   ```bash
   pnpm remove tsup
   pnpm add @vercel/ncc -D
   pnpm add @actions/core -P
   ```

   - We also added the [`@actions/core`](https://github.com/actions/toolkit/tree/master/packages/core) package, used for building GitHub Actions.
     Now we need to update the `build` script in our `package.json`:

   ```diff
   -"build": "tsup",
   +"build": "ncc build src/index.ts -o dist --license licenses.txt",
   ```

   - Our build now emits to the `dist` directory; so we'll want to avoid linting that directory by adding the following to `.eslintignore` and our `.prettierignore`:

   ```diff
   +dist
   ```

   - Rather than having to remember to compile each time, we'll update our precommit hook in `.husky/precommit` to build for us on each commit:

   ```diff
   +pnpm run build
   +git add dist
   npx lint-staged
   ```

> 2. We're going to need an [`action.yml`](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions) file - [here's an example](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action#creating-an-action-metadata-file).

Reviewer: [Proofreading] (re-creating https://github.com/JoshuaKGoldberg/create-typescript-app/pull/1196#discussion_r1443853189) The second linked page links to the first linked page itself. I think we can consolidate this item a bit. How about:

```suggestion
3. Create an [`action.yml` metadata file](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action#creating-an-action-metadata-file).
```

   - update `.github/workflows/build.yml`

     ```diff
     -      - run: node ./lib/index.js
     ```

   - GitHub Actions have a different build mechanism.
     So you'll be removing `tsup` in favour of [`ncc`](https://github.com/vercel/ncc) which compiles output into a single JS file.
     Delete `tsup.config.ts` then execute the following commands:

   ```bash
   pnpm remove tsup
   pnpm add @vercel/ncc -D
   pnpm add @actions/core -P
   ```

   - We also added the [`@actions/core`](https://github.com/actions/toolkit/tree/master/packages/core) package, used for building GitHub Actions.
     Now we need to update the `build` script in our `package.json`:

   ```diff
   -"build": "tsup",
   +"build": "ncc build src/index.ts -o dist --license licenses.txt",
   ```

   - Our build now emits to the `dist` directory; so we'll want to avoid linting that directory by adding the following to `.eslintignore` and our `.prettierignore`:

   ```diff
   +dist
   ```

   - Rather than having to remember to compile each time, we'll update our precommit hook in `.husky/precommit` to build for us on each commit:

   ```diff
   +pnpm run build
   +git add dist
   npx lint-staged
   ```

>    - on the off chance someone isn't running husky, we'll add `.github/workflows/check-dist.yml` ([based on this](https://github.com/actions/typescript-action/blob/main/.github/workflows/check-dist.yml)) to ensure `dist` is up to date:
> 
>    ```yml
>    jobs:
>      check-dist:
>        name: check dist
>        runs-on: ubuntu-latest
> 
>        steps:
>          - uses: actions/checkout@v4
>          - uses: ./.github/actions/prepare
> 
>          - id: build
>            name: Build dist directory
>            run: pnpm run build
> 
>          # This will fail the workflow if the PR wasn't created by Dependabot.
>          - id: diff
>            name: Compare directories
>            run: |
>              if [ "$(git diff --ignore-space-at-eol --text dist/index.js | wc -l)" -gt "0" ]; then
>                echo "Detected uncommitted changes after build."
>                echo "You may need to run 'pnpm run build' locally and commit the changes."
>                echo ""
>                echo "See diff below:"
>                echo ""
>                git diff --ignore-space-at-eol --text dist/index.js
>                echo ""
>                # say this again in case the diff is long
>                echo "You may need to run 'pnpm run build' locally and commit the changes."
>                echo ""
>                exit 1
>              fi
> 
>          # If `dist/` was different than expected, and this was not a Dependabot
>          # PR, upload the expected version as a workflow artifact.
>          - id: upload
>            if: ${{ failure() && steps.diff.outcome == 'failure' }}
>            name: Upload artifact
>            uses: actions/upload-artifact@v4
>            with:
>              name: dist
>              path: dist/
> 
>    name: Check transpiled index.js in dist is up to date
> 
>    on:
>      pull_request: ~
> 
>      push:
>        branches:
>          - main
> 
>    permissions:
>      contents: read
>    ```

Reviewer: This is probably the biggest <code>```suggestion</code> I've ever posted 😄 

```suggestion

    - On the off chance someone isn't running husky, we'll add a workflow to ensure `dist` is up to date:

      <details>
          <summary><code>.github/workflows/dist.yml</code></summary>

      ```yml
      jobs:
        check-dist:
          name: check dist
          runs-on: ubuntu-latest

          steps:
            - uses: actions/checkout@v4
            - uses: ./.github/actions/prepare

            - name: Build dist directory
              run: pnpm run build

            - name: Compare directories
              run: |
                if [ "$(git diff --ignore-space-at-eol --text dist/index.js | wc -l)" -gt "0" ]; then
                  echo "Detected uncommitted changes after build."
                  echo "You may need to run 'pnpm run build' locally and commit the changes."
                  echo ""
                  echo "See diff below:"
                  echo ""
                  git diff --ignore-space-at-eol --text dist/index.js
                  echo ""
                  # say this again in case the diff is long
                  echo "You may need to run 'pnpm run build' locally and commit the changes."
                  echo ""
                  exit 1
                fi

      name: Check built dist/ is up to date

      on:
        pull_request: ~

        push:
          branches:
            - main

      permissions:
        contents: read
      ```

       </details>
```

If you think the tool would be broadly useful to most consumers of this template, feel free to [file a feature request](https://github.com/JoshuaKGoldberg/create-typescript-app/issues/new?assignees=&labels=type%3A+feature&projects=&template=03-feature.yml&title=%F0%9F%9A%80+Feature%3A+%3Cshort+description+of+the+feature%3E) to add it in.

## Can I create a GitHub action with CTA?

> Yes! If you want to read in detail about GitHub Actions [try this](https://docs.github.com/en/actions/creating-actions).

Author: > In HTML, information that is programmatically determinable from a link in English includes text that is in the same paragraph, list, or table cell as the link or in a table header cell that is associated with the table cell that contains the link.

I'd argue that we're satisfying this already: https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context.html#dfn-programmatically-determined-link-context

But I also think that I have no objection to the change and so I commit!

Here we'll outline the steps required to migrate a CTA app to a GitHub Action:

> 1. The GitHub Actions does not cater for tabs well, so you will likely want to set your `README.md` to use spaces, not tabs.

Reviewer: [Question] I know we chatted about this annoyance a bit but I didn't realize it's a blocker. What is it GHA doesn't do well, specifically?

https://github.com/JoshuaKGoldberg/all-contributors-auto-action uses tabs and is fine. Since YML doesn't support tabs _(UGH!)_ the <code>```yml</code> code blocks and `.y(a)ml` files there are spaces still.

Author: The yaml code gets weird wonky highlighting. It's possible that the non yaml text is fine?

Author: Either way, the confusion that arises from mixing tabs and spaces in a single file feels like something worth avoiding.

Reviewer: Do you have an example of the wonkiness? I haven't found anything.

Author: Sure - look at the yaml in the readme of the initial release if Schemar: https://github.com/marketplace/actions/schemar-ci-action?version=v0.1.0

Reviewer: Interesting. That's mixed spaces and tabs - otherwise there'd be red covering all indentations.

Running `pnpm format --write` doesn't change it. But if I manually replace the tabs with spaces, Prettier still is happy. No exit codes either way.

So I think the resolution here should be to not mix spaces and tabs?

Author: Removed the tabs / spaces stuff

> 2. GitHub Actions are _not_ like packages published to npm.
>    The output is expected to live in the `dist` folder of a GitHub repo.

Reviewer: [Proofreading] I think we can convey the same info in fewer words here, thoughts?

```suggestion
2. GitHub Actions store built output on a GitHub branch rather than in a published package on npm.
```

   As a consequence we should:

   - delete `.github/workflows/release.yml` and `.github/workflows/post-release.yml`.
   - update `.github/workflows/build.yml`

>    ```diff
>    -      - run: node ./lib/index.js
>    ```

Reviewer: [Typo] For this and other code blocks, they're not indented enough to be within the intended parent list item.

```suggestion
     ```diff
     -      - run: node ./lib/index.js
     ```
```

>    - GitHub Actions have a different build mechanism.
>      So you'll be removing `tsup` in favour of [`ncc`](https://github.com/vercel/ncc) which compiles output into a single JS file.

Reviewer: [Proofreading] _"a different build mechanism"_ is vague. Someone coming in fresh could be confused by that.

I think what this is trying to convey is that GHAs need all built output in one file rather than dependencies. So how about something like this?

```suggestion
   - GitHub Actions run without installing package dependencies.
     Replace `tsup` with [`ncc`](https://github.com/vercel/ncc) to build source files and dependencies into a single JS file.
```

     Delete `tsup.config.ts` then execute the following commands:

   ```bash
   pnpm remove tsup
   pnpm add @vercel/ncc -D

>    pnpm add @actions/core -P

Reviewer: [Proofreading] I don't think we need to describe setting up `@actions/core` in this guide. Folks will come across it inevitably when learning how to call GitHub APIs. Same with `@actions/github`.

Removing this mention and the _"We also added the ..."_ will get this a bit more terse.

```suggestion
```

   ```

   - We also added the [`@actions/core`](https://github.com/actions/toolkit/tree/master/packages/core) package, used for building GitHub Actions.
     Now we need to update the `build` script in our `package.json`:

   ```diff
   -"build": "tsup",
   +"build": "ncc build src/index.ts -o dist --license licenses.txt",
   ```

   - Our build now emits to the `dist` directory; so we'll want to avoid linting that directory by adding the following to `.eslintignore` and our `.prettierignore`:

   ```diff
   +dist
   ```

   - Rather than having to remember to compile each time, we'll update our precommit hook in `.husky/precommit` to build for us on each commit:

   ```diff
   +pnpm run build
   +git add dist
   npx lint-staged
   ```

>    - on the off chance someone isn't running husky, we'll add `.github/workflows/check-dist.yml` ([based on this](https://github.com/actions/typescript-action/blob/main/.github/workflows/check-dist.yml)) to ensure `dist` is up to date:

Reviewer: [Proofreading] For consistency...

```suggestion
   - On the off chance someone isn't running husky, we'll add `.github/workflows/check-dist.yml` ([based on this](https://github.com/actions/typescript-action/blob/main/.github/workflows/check-dist.yml)) to ensure `dist` is up to date:
```

>    ```yml

Reviewer: [Formatting] Having a big code block in the steps like this takes up a lot of space and makes the steps seem harder than they are. How about using a `<details>` to expando it?

```md

<details>
<summary><code>.github/workflows/dist.yml</code></summary>

...yml

</details>

Author: Sounds interesting - didn't know that was a thing! Do you want to put up a suggestion and I'll commit?

Reviewer: Moved to https://github.com/JoshuaKGoldberg/create-typescript-app/pull/1196/files#r1443870071 👍

>    # In TypeScript actions, `dist/` is a special directory. When you reference
>    # an action with the `uses:` property, `dist/index.js` is the code that will be
>    # run. For this project, the `dist/index.js` file is transpiled from other
>    # source files. This workflow ensures the `dist/index.js` file contains the
>    # expected transpiled code.
>    #
>    # If this workflow is run from a feature branch, it will act as an additional CI
>    # check and fail if the checked-in `dist/` directory does not match what is
>    # expected from the build.

Reviewer: [Proofreading] I've never been a fan of keeping the big explanatory comments from pre-baked templates. They take up a lot of space to document typically introductory level things. Plus they have a tendency to get out of date.

```suggestion
```

   jobs:

>      check-dist:

Reviewer: [Style] Other workflows have kept to minimal names like `lint`. Any thoughts on renaming this to `dist.yml` + `jobs.dist`?

The minimal-ness isn't _quite_ consistent given names like `lint-knip` and `lint-spelling`... but I think once those get consolidated into ESLint (https://github.com/JoshuaKGoldberg/create-typescript-app/milestone/3: #567, #839, #897) we'll be in a good place to rename scripts and their workflows to `knip`, `lint`, `packages`, etc.

Not a blocker at all. 😄

Author: Yeah I guess dist.yml would be fine... An alternative idea would be to combine the dist check with the `build.yml` - thoughts?

Reviewer: Oh I like that!

Which reminds me - I think there should be _two_ build task:

* Release builds: which is `pnpm build:release` in my repos
* Type checking: which is `pnpm build` in my repos

Author: I'll move the `check-dist` stuff into `build` shortly.

> Release builds: which is pnpm build:release in my repos
> Type checking: which is pnpm build in my repos

I find this naming slightly confusing. The question it prompts to me is "I have `build` and `build:release` - are they building different things? Does it matter?"  If it was just `release` and not `build:release` I think I'd think "okay, that one's pushing out to npm".  But `build:release` makes me do a double take

Reviewer: Hmm yeah. Maybe it should be `"build" "ncc ..."` and `"tsc": "tsc"`? Thoughts?

Author: I actually don't love `tsc` as a name. Questions it prompts: "tsc can output js and typecheck, I wonder what tsc means here?" Ideally names (for my money) should prompt as few questions as possible. How about `typecheck`? (As I think tsc is just being used for type checking?)

Reviewer: Blurgh there's something about `typecheck` that saddens me as a name... All the other ones that don't have a `:` are one word. Then this is two. And so long!

Agreed with the gripes about `tsc`. I don't love it either. 😦

       name: check dist
       runs-on: ubuntu-latest

       steps:
         - uses: actions/checkout@v4
         - uses: ./.github/actions/prepare

         - id: build
           name: Build dist directory
           run: pnpm run build

>          # This will fail the workflow if the PR wasn't created by Dependabot.

Reviewer: > Dependabot

This template uses Renovate. But also I don't see why they're relevant?

         - id: diff
           name: Compare directories
           run: |

>              if [ "$(git diff --ignore-space-at-eol --text dist/index.js | wc -l)" -gt "0" ]; then

Reviewer: [Question] https://github.com/tj-actions/verify-changed-files does the same thing. Why not use it?

This can be a followup if you don't want to investigate a whole new action approach for a docs PR 😄 (I wouldn't)

Author: yeah I definitely don't :smile: - but it sounds cool

               echo "Detected uncommitted changes after build."
               echo "You may need to run 'pnpm run build' locally and commit the changes."
               echo ""
               echo "See diff below:"
               echo ""
               git diff --ignore-space-at-eol --text dist/index.js
               echo ""
               # say this again in case the diff is long
               echo "You may need to run 'pnpm run build' locally and commit the changes."
               echo ""
               exit 1
             fi

> 
>          # If `dist/` was different than expected, and this was not a Dependabot
>          # PR, upload the expected version as a workflow artifact.
>          - id: upload
>            if: ${{ failure() && steps.diff.outcome == 'failure' }}
>            name: Upload artifact
>            uses: actions/upload-artifact@v4
>            with:
>              name: dist
>              path: dist/

Reviewer: [Question] Why? Who needs this artifact?

```suggestion
```

Author: Good question - I don't know.  I can remove it?  It came from where I borrowed it in the first place - quite possibly not required

Reviewer: Yeah I say let's remove it & discover any gaps the hard way.

Author: Yeah agreed - I'll remove. Can probably get to this tonight if all goes smoothly

   name: Check transpiled index.js in dist is up to date

   on:
     pull_request: ~

     push:
       branches:

>          - main
>    permissions:

Reviewer: [Style] Nitty nit, for blank line consistency...

```suggestion
         - main

   permissions:
```

     contents: read
   ```

> 3. We're going to need an [`action.yml`](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions) file - [here's an example](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action#creating-an-action-metadata-file).

Reviewer: [Proofreading] The second linked page links to the first linked page itself. I think we can consolidate this item a bit. How about:

```suggestion
3. Create an [`action.yml` metadata file](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action#creating-an-action-metadata-file).
```

Author: Sounds good - I want to hit commit suggestion but I think you made it in an old commit

Reviewer: Heh silly GitHub - I'll recreate

> 4. Our GitHub Action needs Node.js 20 so we'll update `.github/actions/prepare/action.yml`:
> 
> ```diff
> -node-version: "18"
> +node-version: "20"
> ```

Reviewer: [Question] Is Node.js 20 needed? My actions have been running on 18 just fine.

```suggestion
```

Also filed #1204 to bump to 20 for everyone.

Author: yeah I think it is - or at least I read it was

> 
> 5. Change the code in your `src` directory to be a GitHub Action.
>    It's worth reading the official documentation [for an example](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action#writing-the-action-code).

Reviewer: [Proofreading] I think changing the code in `src` to be an action is implied 😄. IMO adding another step to mention it makes the steps seem longer without any user benefit.

Though I do like the mention of the GHA docs... how about:

```suggestion

It's worth reading the [GitHub Actions documentation](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action#writing-the-action-code).
```

       push:
         branches:
           - main

     permissions:
       contents: read
     ```

      </details>

   - GitHub Actions run without installing package dependencies.
     Replace `tsup` with [`ncc`](https://github.com/vercel/ncc) to build source files and dependencies into a single JS file.
     Delete `tsup.config.ts` then execute the following commands:

   ```bash
   pnpm remove tsup
   pnpm add @vercel/ncc -D
   ```

   - Now we need to update the `build` script in our `package.json`:

   ```diff
   -"build": "tsup",
   +"build": "ncc build src/index.ts -o dist --license licenses.txt",
   ```

   - Our build now emits to the `dist` directory; so we'll want to avoid linting that directory by adding the following to `.eslintignore` and our `.prettierignore`:

   ```diff
   +dist
   ```

   - Rather than having to remember to compile each time, we'll update our precommit hook in `.husky/precommit` to build for us on each commit:

   ```diff
   +pnpm run build
   +git add dist
   npx lint-staged
   ```

2. Create an [`action.yml` metadata file](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action#creating-an-action-metadata-file).

> 3. Our GitHub Action needs Node.js 20 so we'll update `.github/actions/prepare/action.yml`:
> 
> ```diff
> -node-version: "18"
> +node-version: "20"
> ```

Reviewer: #1233 bumped Node to 20, so this step is no longer necessary.

```suggestion
```

I don't trust GitHub to get this suggestion out properly so I'll just apply it. 😄
